plugins/api/src/switch_user.cpp

```diff
@@ -0,0 +1,170 @@
+#include "irods/plugins/api/api_plugin_number.h"
+#include "irods/plugins/api/switch_user_types.h"
+#include "irods/rodsDef.h"
+#include "irods/rcConnect.h"
+#include "irods/rodsPackInstruct.h"
+#include "irods/rcMisc.h"
+#include "irods/client_api_allowlist.hpp"
+
+#include "irods/apiHandler.hpp"
+
+#include <functional>
+
+#ifdef RODS_SERVER
+
+//
+// Server-side Implementation
+//
+
+#  include "irods/irods_rs_comm_query.hpp"
+#  include "irods/rodsErrorTable.h"
+#  include "irods/irods_logger.hpp"
+#  include "irods/server_utilities.hpp"
+
+#  define IRODS_USER_ADMINISTRATION_ENABLE_SERVER_SIDE_API
+#  include "irods/user_administration.hpp"
+
+#  include <cstring>
+#  include <string>
+#  include <string_view>
+
+namespace
+{
+    //
+    // Function Prototypes
+    //
+
+    auto call_switch_user(irods::api_entry*, RsComm*, SwitchUserInput*) -> int;
+    auto rs_switch_user(RsComm*, SwitchUserInput*) -> int;
+
+    //
+    // Function Implementations
+    //
+
+    auto call_switch_user(irods::api_entry* _api, RsComm* _comm, SwitchUserInput* _input)
+        -> int
+    {
+        return _api->call_handler<SwitchUserInput*>(_comm, _input);
+    } // call_switch_user
+
+    auto rs_switch_user(RsComm* _comm, SwitchUserInput* _input) -> int
+    {
+        using log_api = irods::experimental::log::api;
+
+        // Only administrators are allowed to invoke this API.
+        // We check the proxy user because it covers clients and server-to-server redirects.
+        // Remember, the client and proxy user information is identical until there's a redirect.
+        if (!irods::is_privileged_proxy(*_comm)) {
+            log_api::error(
+                "Proxy user [{}] does not have permission to switch client user.", _comm->proxyUser.userName);
+            return SYS_PROXYUSER_NO_PRIV;
+        }
+
+        if (!_input) {
+            log_api::error("Invalid input argument: Received a null pointer");
+            return SYS_INVALID_INPUT_PARAM;
+        }
+
+        const auto* local_zone = getLocalZoneName();
+
+        if (!local_zone) {
+            log_api::error("Could not get name of local zone.");
+            return SYS_INTERNAL_ERR;
+        }
+
+        const auto* username = _input->username;
+        const auto* zone = _input->zone;
+        bool is_local_zone = false;
+
+        // Passing an empty string as the zone argument is equivalent to explicitly setting
+        // the zone parameter to the local zone.
+        if (const std::string_view z{_input->zone}; z.empty() || z == local_zone) {
+            zone = local_zone;
```

Comment:
I'm thinking of changing the API so that it forces users to pass a non-empty zone.

The optimization where we check the username and zone to avoid unnecessary network interaction assumes the client app always passes a non-empty zone, yet the API plugin allows clients to pass empty zone strings.

The issue is that the rc_switch_user function copies the client-provided zone argument into the zone member variable of the RcComm following success. If the zone argument was empty, the client will now be in a situation.

Therefore, we can't accept empty zone names.

Thoughts?

Reply:
The more I think about this, the more I feel requiring the client to pass a zone is the correct thing to do.

If we don't require the zone, clients can simply provide a username and let the server become whoever it wants.

Reply:
this seems very reasonable.   there is no legacy code to worry about / support.

no way to hold it wrong is a good thing.